README.md
# Jet Reclustering

*This is currently in AnalysisBase releases 2.4.17+.*

This tool allows you to recluster small-R xAOD jets into large-R xAOD jets. It provides configurable filtering of the small-R jets, reclustering using standard or variable-R algorithms, configurable trimming of the large-R jets, and jet moment & jet substructure moment calculations.

If you would like to get involved, see the twiki for [the JetMET working group for jet reclustering](https://twiki.cern.ch/twiki/bin/view/AtlasProtected/JetReclustering). The [pre-recommendations](https://twiki.cern.ch/twiki/bin/viewauth/AtlasProtected/PreRec) twiki contains the guidelines for your analyses.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Installing](#installing)
- [Configurations for](#configurations-for)
  - [`JetReclusteringTool` tool](#jetreclusteringtool-tool)
  - [`JetReclusteringAlgo` algorithm](#jetreclusteringalgo-algorithm)
  - [`AthJetReclusteringAlgo` Athena algorithm](#athjetreclusteringalgo-athena-algorithm)
- [Using Jet Reclustering](#using-jet-reclustering)
  - [Input Jet Filtering](#input-jet-filtering)
  - [Output Reclustered Jet Trimming](#output-reclustered-jet-trimming)
  - [Variable-R Jet Finding](#variable-r-jet-finding)
  - [Area Calculations](#area-calculations)
  - [Incorporating in existing code](#incorporating-in-existing-code)
    - [RootCore](#rootcore)
    - [Athena](#athena)
  - [Incorporating in algorithm chain](#incorporating-in-algorithm-chain)
    - [RootCore](#rootcore-1)
    - [Athena](#athena-1)
- [Studies and Example Usage](#studies-and-example-usage)
  - [Accessing the subjets from constituents](#accessing-the-subjets-from-constituents)
  - [Accessing various jet moments](#accessing-various-jet-moments)
- [Authors](#authors)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Installing

This works in AB 2.3.X and 2.4.Y on ROOT6 releases. As long as [JetRec](http://acode-browser.usatlas.bnl.gov/lxr/source/atlas/Reconstruction/Jet/JetRec/JetRec/) works, this will be ok.

```bash
rcSetup Base,2.4.Y # or 2.3.X
git clone https://github.com/kratsg/JetReclustering.git
rc find_packages
rc compile
```

*Warning*: use `X >= 45` since I am using `ANA_CHECK` which only works with `EventLoop-00-01-35` or better. Use `Y >= 17` as it is part of the atlas software releases.

## Configurations for

### `JetReclusteringTool` tool

 Property           | Type                      | Default                   | Description
:-------------------|:-------------------------:|--------------------------:|:-------------------------------------------------------------------------------------
InputJetContainer   | string                    |                           | name of the input jet container for reclustering
OutputJetContainer  | string                    |                           | name of the output jet container holding reclustered jets
InputJetPtMin       | float                     | 25.0                      | filter input jets by requiring a minimum pt cut [GeV]
ReclusterAlgorithm  | string                    | AntiKt                    | name of algorithm for clustering large-R jets {AntiKt, Kt, CamKt}
ReclusterRadius     | float                     | 1.0                       | radius of large-R reclustered jets or maximum radius of variable-R jet finding
RCJetPtMin          | float                     | 50.0                      | filter reclustered jets by requiring a minimum pt cut [GeV]
RCJetPtFrac         | float                     | 0.05                      | trim the reclustered jets with a PtFrac on its constituents (eg: small-R input jets)
RCJetSubjetRadius   | float                     | 0.0                       | radius parameter for kt-clustering to form subjets (R=0.0 should not do any clustering)
VariableRMinRadius  | float                     | -1.0                      | minimum radius for variable-R jet finding
VariableRMassScale  | float                     | -1.0                      | mass scale [GeV] for variable-R jet finding
DoArea              | bool                      | false                     | turn on ghost area calculations (set ghost area scale to 0.01)
AreaAttributes      | string                    | ActiveArea ActiveArea4vec | space-delimited list of attributes to transfer over from fastjet

### `JetReclusteringAlgo` algorithm

As well as the provided above configurations for the `JetReclusteringTool`, we also provide a `m_debug` configuration for extra verbose output and an `m_outputXAODName` to create an output xAOD containing the reclustered jets (note: experimental)

Variable            | Type      | Default                   | Description
:-------------------|:---------:|--------------------------:|:-------------------------------------------------------------------------------------
m_inputJetContainer | string    |                           | see above
m_outputJetContainer| string    |                           | see above
m_ptMin_input       | float     | 25.0                      | see above
m_rc_alg            | string    | AntiKt                    | see above
m_radius            | float     | 1.0                       | see above
m_ptMin_rc          | float     | 50.0                      | see above
m_ptFrac            | float     | 0.05                      | see above
m_subjet_radius     | float     | 0.0                       | see above
m_varR_minR         | float     | -1.0                      | see above
m_varR_mass         | float     | -1.0                      | see above
m_doArea            | bool      | false                     | see above
m_areaAttributes    | string    | ActiveArea ActiveArea4vec | see above
m_outputXAODName    | string    |                           | if defined, put the reclustered jets in an output xAOD file of the given name
m_debug             | bool      | false                     | enable verbose debugging information, such as printing the tool configurations

### `AthJetReclusteringAlgo` Athena algorithm

 Property           | Type                      | Default                   | Description
:-------------------|:-------------------------:|--------------------------:|:-------------------------------------------------------------------------------------
JetReclusteringTool          | ToolHandle                |                           | The JetReclusteringTool to use. All configurables should be set on this.

## Using Jet Reclustering

### Input Jet Filtering

The input jets can be filtered using the `InputJetPtMin` or `m_ptMin_input` options. If these are set to 0 (or less), this will turn the jet filtering tool off (essentially skipping the step).

### Output Reclustered Jet Trimming

The output jets can be trimmed using the `RCJetPtFrac` or `m_ptFrac` options. If these are set to 0 (or less), this will turn the reclustered jet trimming tool off (essentially skipping the step). Note that by default, we will not kt-cluster the subjets as `RCJetSubjetRadius` / `m_subjet_radius` are set to 0.0 by default, however one can set them to a non-zero radius value if you want to form subjets (if you somehow passed in `xAOD::Jet` objects that represent clusters rather than small-R jets).

### Variable-R Jet Finding

Variable-R jet finding is performed if `VariableRMinRadius >= 0` and `VariableRMassScale >= 0`. For more information on these variables, see the [Jets with Variable R](http://arxiv.org/pdf/0903.0392v1.pdf) paper. If you choose variable-R jet finding, the maximum jet radius will be specified by `ReclusterRadius`. The relevant properties are listed in the following table

 Property           | Type                      | Default                   | Description
:-------------------|:-------------------------:|--------------------------:|:-------------------------------------------------------------------------------------
ReclusterRadius     | float                     | 1.0                       | maximum radius of variable-R jet finding
VariableRMinRadius  | float                     | -1.0                      | minimum radius for variable-R jet finding
VariableRMassScale  | float                     | -1.0                      | mass scale [GeV] for variable-R jet finding

When a new jet is formed using variable-R jet finding, it will have some extra attributes as mentioned on the [Run 2 - Jet Moments](https://twiki.cern.ch/twiki/bin/viewauth/AtlasProtected/Run2JetMoments) twiki. We will also decorate with an `EffectiveR` attribute as well which reflects the effective radius of the reclustered jet using its untrimmed transverse momentum. To summarize the translations

 Property           | Type                      | Jet Attribute
:-------------------|:-------------------------:|--------------------
ReclusterRadius     | float                     | SizeParameter
VariableRMinRadius  | float                     | VariableRMinRadius
VariableRMassScale  | float                     | VariableRMassScale
                    | float                     | EffectiveR

### Area Calculations

Areas can be calculated and added to the jets. Fastjet does the area calculation and these values can be transferred over using the `JetFromPseudojet` tool. To make this happen, simply enable `m_doArea` (`DoArea`) which will set the ghost area size to `0.01` which is a reasonable default for most use cases. The attributes that get transferred over are defined in `m_areaAttributes` (`AreaAttributes`). As of the time of writing this section of the README, there were only two allowed values which would get decorated:

- `ActiveArea` (most people use this one)
- `ActiveArea4vec`

### Incorporating in existing code

#### RootCore

If you wish to incorporate `JetReclustering` directly into your code, add this package as a dependency in `cmt/Makefile.RootCore` and then a header

```c++
#include <AsgTools/AnaToolHandle.h>
#include <JetInterface/IJetExecuteTool.h>

class MyAlgo : public EL::Algorithm {
  // ...

  asg::AnaToolHandle<IJetExecuteTool> m_jetReclusteringTool; //!
}
```

to get started. In the source, you need to add the tool header

```c++
#include <JetReclustering/JetReclusteringTool.h>
```

then make sure the AsgTool tool store sets up the tool correctly in the constructor

```c++
MyAlgo :: MyAlgo () :
  m_jetReclusteringTool("JetReclusteringTool/ANameForTheTool")
  {}
```

At this point, you can set up your standard tool in the `initialize()` portion of your algorithm as a tool handle

```c++
ANA_CHECK_SET_TYPE (EL::StatusCode);
ANA_CHECK(ASG_MAKE_ANA_TOOL(m_jetReclusteringTool, JetReclusteringTool));
ANA_CHECK(m_jetReclusteringTool.setProperty("InputJetContainer",  m_inputJetContainer));
ANA_CHECK(m_jetReclusteringTool.setProperty("OutputJetContainer", m_outputJetContainer));
ANA_CHECK(m_jetReclusteringTool.setProperty("ReclusterRadius",    m_radius));
ANA_CHECK(m_jetReclusteringTool.setProperty("ReclusterAlgorithm", m_rc_alg));
ANA_CHECK(m_jetReclusteringTool.setProperty("VariableRMinRadius", m_varR_minR));
ANA_CHECK(m_jetReclusteringTool.setProperty("VariableRMassScale", m_varR_mass));
ANA_CHECK(m_jetReclusteringTool.setProperty("InputJetPtMin",      m_ptMin_input));
ANA_CHECK(m_jetReclusteringTool.setProperty("RCJetPtMin",         m_ptMin_rc));
ANA_CHECK(m_jetReclusteringTool.setProperty("RCJetPtFrac",        m_ptFrac));
ANA_CHECK(m_jetReclusteringTool.setProperty("RCJetSubjetRadius",  m_subjet_radius));
ANA_CHECK(m_jetReclusteringTool.setProperty("DoArea",             m_doArea));
ANA_CHECK(m_jetReclusteringTool.setProperty("AreaAttributes",     m_areaAttributes));
ANA_CHECK(m_jetReclusteringTool.retrieve());
```

and then simply call `m_jetReclusteringTool->execute()` in the `execute()` portion of your algorithm to fill the TStore with the appropriate container(s). Note that you use a pointer on the second portion when calling `execute()` to access the underlying pointer to the tool itself. The functions `setProperty()` and `initialize()` have a return type `StatusCode` which needs to be checked.

#### Athena

The methods for incorporating the code into an Athena algorithm are very similar to the method for RootCore.
The package must be included with a 'use' statement in `cmt/requirements`. Then in your header:
```c++
#include <AsgTools/ToolHandle.h> // Can use AnaToolHandle instead if you want
#include <JetInterface/IJetExecuteTool.h>

class MyAlgo : public ::AthAnalysisAlgorithm { // Or any of the other Athena algorithm types
  // ...

  ToolHandle<IJetExecuteTool> m_jetReclusteringTool;
};
```

As with other Athena algorithms it's recommended to configure the tool in your job options and then pass it into your algorithm as a property. To do this, in the constructor you need to initialize the handle and declare it as a property

```c++

MyAlgo::MyAlgo( const std::string& name, ISvcLocator* pSvcLocator )
: AthAnalysisAlgorithm( name, pSvcLocator ), // Or other base class
  m_jetReclusteringTool("") // Can also initialize with a default type/name if you wish
{
  // Other properties...
  declareProperty( "JetReclusteringTool", m_jetReclusteringTool );
}
```

Then in the `initialize()` member function of your algorithm `retrieve` the tool

```c++
  ATH_CHECK( m_jetReclusteringTool.retrieve() );
```

To use it in your algorithm you can call (in your `execute()` function). Make sure you do this **before** anyone tries to use the output...
```c++
  int retCode = m_jetReclusteringTool->execute();
  if (retCode != 0) {
    ATH_MSG_ERROR( "JetReclusteringTool failed in execution with code: " << retCode );
    return StatusCode::FAILURE;
  }
```

Then you need to create and pass in your tool in your joboptions (see the section below for more instructions, replace `AthJetReclusteringAlgo` with `MyAlgo` or whatever you called your algorithm).
It is also possible to set the properties and initialize your tool in the cxx code but this isn't really recommended. If you want to do this either use AnaToolHandle and setProperty (analogously to in RootCore) or use ToolHandle and [AthAnalysisHelper](http://acode-browser2.usatlas.bnl.gov/lxr-AthAna/source/atlas/Control/AthAnalysisBaseComps/AthAnalysisBaseComps/AthAnalysisHelper.h#0053).

### Incorporating in algorithm chain

#### RootCore

This is the least destructive option since it requires **no change** to your existing code. All you need to do is create a new `JetReclusteringAlgo` algorithm and add it to the job before other algorithms downstream that want access to the reclustered jets. It is highly configurable. In your runner macro, add the header

```c++
#include <JetReclustering/JetReclusteringAlgo.h>
```

and then simply set up your algorithm like so

```c++
// initialize and set it up
JetReclustering* jetReclusterer = new JetReclusteringAlgo();
jetReclusterer->m_inputJetContainer = "AntiKt4LCTopoJets";
jetReclusterer->m_outputJetContainer = "AntiKt10LCTopoJetsRCAntiKt4LCTopoJets";
jetReclusterer->m_name = "R10"; // unique name for the tool
jetReclusterer->m_ptMin_input = 25.0; // GeV
jetReclusterer->m_ptMin_rc = 50.0; // GeV
jetReclusterer->m_ptFrac = 0.05; // GeV

// ...
// ...
// ...

// add it to your job sometime later
job.algsAdd(jetReclusterer);
```

#### Athena

As with RootCore this is the best way to use the tool (Athena is designed to execute a series of algorithms after all...).
All you need to do is create a `JetReclusteringTool` in the `ToolSvc`, add an `AthJetReclusteringAlgo` to your `AlgSequence` and connect them

```python
ToolSvc += CfgMgr.JetReclusteringTool("MyJetReclusteringTool", InputJetContainer = "AntiKt4EMTopoJets", OutputJetContainer = "AntiKt10EMTopoJets_RC") # Set up properties here
ToolSvc.MyJetReclusteringTool.InputJetPtMin = 10 #Can also set properties like this

algseq = CfgMgr.AthSequencer("AthAlgSeq")                #gets the main AthSequencer
algseq += CfgMgr.AthJetReclusteringAlgo("JetRecAlgo", JetReclusteringTool = ToolSvc.MyJetReclusteringTool)
```

Now any algorithm downstream of this in algseq will have access to the `AntiKt10EMPTopoJets_RC` container.
An example (and a test you can use) is in `share/AthJetReclusteringAlgoJobOptions.py`

## Studies and Example Usage

See [kratsg/ReclusteringStudies](https://github.com/kratsg/ReclusteringStudies) for studies and example usage.

### Accessing the subjets from constituents

The reclustered jets have constituents which are your input small-R jets. These can be re-inflated, so to speak. As an example, I wanted to get the btagging information of my subjets as well as their constituents (eg: the topological calorimeter clusters, `TopoCaloClusters`)

```c++
for(auto jet: *in_jets){
  const xAOD::Jet* subjet(nullptr);
  const xAOD::BTagging* btag(nullptr);
  for(auto constit: jet->getConstituents()){
    if (subjet->type() != xAOD::Type::Jet) {
      // You need to handle this somehow
      continue;
    }
    subjet = static_cast<const xAOD::Jet*>(constit->rawConstituent());
    btag = subjet->btagging();
    if(btag)
      Info("execute()", "btagging: %0.2f", btag->MV1_discriminant());

    for(auto subjet_constit: subjet->getConstituents())
      Info("execute()", "\tconstituent pt: %0.2f", subjet_constit->pt());
  }
}
```

where we explicitly `static_cast<>` our raw pointer from the `rawConstituent()` call. See [xAODJet/JetConstituentVector.h](http://acode-browser.usatlas.bnl.gov/lxr/source/atlas/Event/xAOD/xAODJet/xAODJet/JetConstituentVector.h) for more information about what is available. As a raw pointer, we already know that the input to the constituents were small-R jets (since we re-clustered ourselves) so this type of casting is safe.

### Accessing various jet moments

We try to use the standard `JetModifier` tools that are available ATLAS-wide. In those cases, you can find a lot more information on the [Run 2 - Jet Moments](https://twiki.cern.ch/twiki/bin/viewauth/AtlasProtected/Run2JetMoments) twiki. There are two ways to figure out what information is stored on the reclustered jet.


ChangeLog
2016-08-23 Will Buttinger <will@cern.ch>
	* test/JetReclustering.xml: fix so that atn is happy with it
	* test/ut_JetReclusteringTool_test.cxx : extend with complete example


Root/JetReclusteringTool.cxx
#include "JetReclustering/JetReclusteringTool.h"
#include "AsgTools/Check.h"

#include "JetInterface/IJetModifier.h"
#include "JetInterface/IJetFromPseudojet.h"
#include "JetInterface/IJetGroomer.h"
#include "JetInterface/IPseudoJetGetter.h"
#include "JetInterface/IJetFinder.h"
#include "JetInterface/IJetExecuteTool.h"

#include <sstream>

#ifdef ROOTCORE // The macro for RootCore needs these (provides a way of getting around missing dictionaries)
#include "JetRec/JetFilterTool.h"
#include "JetRec/JetRecTool.h"
#include "JetRec/PseudoJetGetter.h"
#include "JetRec/JetFromPseudojet.h"
#include "JetRec/JetFinder.h"
#include "JetRec/JetTrimmer.h"
#include "JetReclustering/EffectiveRTool.h"
#include "JetSubStructureMomentTools/JetChargeTool.h"
#include "JetSubStructureMomentTools/JetPullTool.h"
#include "JetSubStructureMomentTools/EnergyCorrelatorTool.h"
#include "JetSubStructureMomentTools/EnergyCorrelatorRatiosTool.h"
#include "JetSubStructureMomentTools/KTSplittingScaleTool.h"
#include "JetSubStructureMomentTools/DipolarityTool.h"
#include "JetSubStructureMomentTools/CenterOfMassShapesTool.h"
#include "JetSubStructureMomentTools/NSubjettinessTool.h"
#endif

JetReclusteringTool::JetReclusteringTool(std::string name) :
  AsgTool(name),
  m_jetFilterTool("JetFilterTool/JetFilterTool_" + this->name()),
  m_inputJetFilterTool("JetRecTool/JetRec_InputJetFilterTool_" + this->name()),
  m_pseudoJetGetterTool("PseudoJetGetter/PseudoJetGetterTool_" + this->name()),
  m_jetFromPseudoJetTool("JetFromPseudojet/JetFromPseudoJetTool_" + this->name()),
  m_jetFinderTool("JetFinder/JetFinderTool_" + this->name()),
  m_reclusterJetTool("JetRecTool/JetRec_JetReclusterTool_" + this->name()),
  m_trimJetTool("JetRecTool/JetRec_JetTrimTool_" + this->name()),
  m_effectiveRTool("EffectiveRTool/EffectiveRTool"),
  m_jetTrimmingTool("JetTrimmer/JetTrimmerTool_" + this->name()),
  m_jetChargeTool("JetChargeTool/JetChargeTool_" + this->name()),
  m_jetPullTool("JetPullTool/JetPullTool_" + this->name()),
  m_energyCorrelatorTool("EnergyCorrelatorTool/EnergyCorrelatorTool_" + this->name()),
  m_energyCorrelatorRatiosTool("EnergyCorrelatorRatiosTool/EnergyCorrelatorRatiosTool_" + this->name()),
  m_ktSplittingScaleTool("KTSplittingScaleTool/KTSplittingScaleTool_" + this->name()),
  m_dipolarityTool("DipolarityTool/DipolarityTool_" + this->name()),
  m_centerOfMassShapesTool("CenterOfMassShapesTool/CenterOfMassShapesTool_" + this->name()),
  m_nSubjettinessTool("NSubjettinessTool/NSubjettinessTool_" + this->name())
{


  declareProperty("InputJetContainer",  m_inputJetContainer = "");
  declareProperty("OutputJetContainer", m_outputJetContainer = "");
  declareProperty("ReclusterRadius",    m_radius = 1.0);
  declareProperty("ReclusterAlgorithm", m_rc_alg = "AntiKt");
  declareProperty("VariableRMinRadius", m_varR_minR =-1.0);
  declareProperty("VariableRMassScale", m_varR_mass =-1.0);
  declareProperty("InputJetPtMin",      m_ptMin_input = 25.0);
  declareProperty("RCJetPtMin",         m_ptMin_rc = 50.0);
  declareProperty("RCJetPtFrac",        m_ptFrac = 0.05);
  declareProperty("RCJetSubjetRadius",  m_subjet_radius = 0.2);
  declareProperty("DoArea",             m_doArea = false);
  declareProperty("AreaAttributes",     m_areaAttributes = "ActiveArea ActiveArea4vec");

}

StatusCode JetReclusteringTool::initialize(){
  if(m_isInitialized){
    ATH_MSG_WARNING(m_APP_NAME << " already initialized.");
    return StatusCode::FAILURE;
  }
  ATH_MSG_INFO(m_APP_NAME << " initializing");
  ATH_MSG_INFO( "Name is " << name());


  ATH_MSG_INFO( "... with containers:" );
  ATH_MSG_INFO( "\tInputJetContainer: " << m_inputJetContainer );
  ATH_MSG_INFO( "\tOutputJetContainer: " << m_outputJetContainer );

  // set to true, we're calling it now
  m_isInitialized = true;

  // set up some stuff to use within the initialize()
  ToolHandleArray<IJetModifier> modArray;
  ToolHandleArray<IPseudoJetGetter> getterArray;

  // this is for intermediate output containers used between tools
  std::string filteredInputJetContainer(m_inputJetContainer);
  if(m_ptMin_input > 0) filteredInputJetContainer = "Filtered_"+name();
  std::string filteredInputPseudoJetsContainer = "PseudoJets_"+name();
  std::string reclusteredJetsContainer = "ReclusteredJets_"+name();

  if(m_ptMin_input > 0){
    ATH_MSG_INFO( "Input Jet Filtering" );
    ATH_MSG_INFO( "\tPtMin: " << m_ptMin_input*1.e3 << " GeV" );
    /* initialize input jet filtering */
    //    - create a tool that will filter jets
    ASG_CHECK( ASG_MAKE_ANA_TOOL( m_jetFilterTool, JetFilterTool) );
    ASG_CHECK(m_jetFilterTool.setProperty("PtMin", m_ptMin_input*1.e3));
    ASG_CHECK(m_jetFilterTool.setProperty("OutputLevel", msg().level() ) );
    ASG_CHECK(m_jetFilterTool.retrieve() );
    modArray.clear();
    modArray.push_back(m_jetFilterTool.handle() );
    ATH_CHECK(modArray.retrieve() );

    //    - create the master tool to filter the input jets
    ASG_CHECK( ASG_MAKE_ANA_TOOL( m_inputJetFilterTool, JetRecTool) );
    ASG_CHECK(m_inputJetFilterTool.setProperty("InputContainer", m_inputJetContainer));
    ASG_CHECK(m_inputJetFilterTool.setProperty("OutputContainer", filteredInputJetContainer));
    ASG_CHECK(m_inputJetFilterTool.setProperty("JetModifiers", modArray));
    /* note: we cannot use shallow copies since we are removing elements from a
     * container, we need a deep copy as linking will break */
    ASG_CHECK(m_inputJetFilterTool.setProperty("ShallowCopy", false));
    ASG_CHECK(m_inputJetFilterTool.setProperty("OutputLevel", msg().level() ) );
    ASG_CHECK(m_inputJetFilterTool.retrieve());
  }

  // only compute area if ptFrac = 0.0 and m_areaAttributes is specified
  float ghostArea(0.0);
  std::vector<std::string> areaAttributes;
  if(m_doArea){
    ATH_MSG_INFO( "Ghost Area Calculations:" );
    ATH_MSG_INFO( "\tAreaAttributes: " << m_areaAttributes );
    ghostArea = 0.01;
    // split up the m_areaAttributes string specifying which attributes to record
    std::string token;
    std::istringstream ss(m_areaAttributes);
    while(std::getline(ss, token, ' '))
      areaAttributes.push_back(token);
  }
  /* initialize jet reclustering */
  //    - create a PseudoJet builder.
  ATH_MSG_INFO( "PseudoJet Builder initializing..." );
  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_pseudoJetGetterTool, PseudoJetGetter) );
  ASG_CHECK(m_pseudoJetGetterTool.setProperty("InputContainer", filteredInputJetContainer));
  ASG_CHECK(m_pseudoJetGetterTool.setProperty("OutputContainer", filteredInputPseudoJetsContainer));
  ASG_CHECK(m_pseudoJetGetterTool.setProperty("Label", "LCTopo"));
  ASG_CHECK(m_pseudoJetGetterTool.setProperty("SkipNegativeEnergy", true));
  ASG_CHECK(m_pseudoJetGetterTool.setProperty("GhostScale", 0.0));
  ASG_CHECK(m_pseudoJetGetterTool.setProperty("OutputLevel", msg().level() ) );
  ASG_CHECK(m_pseudoJetGetterTool.retrieve());
  getterArray.push_back(m_pseudoJetGetterTool.handle());
  ATH_CHECK(getterArray.retrieve() );
  //    - create a Jet builder
  ATH_MSG_INFO( "Jet Builder initializing..." );
  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_jetFromPseudoJetTool, JetFromPseudojet) );
  ASG_CHECK(m_jetFromPseudoJetTool.setProperty("Attributes", areaAttributes));
  ASG_CHECK(m_jetFromPseudoJetTool.setProperty("OutputLevel", msg().level() ) );
  ASG_CHECK(m_jetFromPseudoJetTool.retrieve());
  //    - create a ClusterSequence Tool
  ATH_MSG_INFO( "Cluster Sequencer initializing..." );
  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_jetFinderTool, JetFinder) );
  ASG_CHECK(m_jetFinderTool.setProperty("JetAlgorithm", m_rc_alg));
  ASG_CHECK(m_jetFinderTool.setProperty("JetRadius", m_radius));
  ASG_CHECK(m_jetFinderTool.setProperty("VariableRMinRadius", m_varR_minR));
  ASG_CHECK(m_jetFinderTool.setProperty("VariableRMassScale", m_varR_mass*1.e3));
  ASG_CHECK(m_jetFinderTool.setProperty("PtMin", m_ptMin_rc*1.e3));
  // set ghost area, ignore if trimming is being applied to reclustered jets
  ASG_CHECK(m_jetFinderTool.setProperty("GhostArea", ghostArea));
  ASG_CHECK(m_jetFinderTool.setProperty("RandomOption", 1));
  ASG_CHECK(m_jetFinderTool.setProperty("JetBuilder", m_jetFromPseudoJetTool.handle()));
  ASG_CHECK(m_jetFinderTool.setProperty("OutputLevel", msg().level() ) );
  ASG_CHECK(m_jetFinderTool.retrieve());
  //    - create list of modifiers.
  modArray.clear();
  //        we need to calculate effectiveR before trimming, if we are doing variableR
  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_effectiveRTool, EffectiveRTool) );
  ATH_CHECK(m_effectiveRTool.retrieve());
  modArray.push_back(m_effectiveRTool.handle() );
  ATH_CHECK(modArray.retrieve() );
  //    - create our master reclustering tool
  ATH_MSG_INFO( "Jet Reclusterer initializing..." );
  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_reclusterJetTool, JetRecTool) );
  ASG_CHECK(m_reclusterJetTool.setProperty("OutputContainer", reclusteredJetsContainer));
  ASG_CHECK(m_reclusterJetTool.setProperty("PseudoJetGetters", getterArray));
  ASG_CHECK(m_reclusterJetTool.setProperty("JetFinder", m_jetFinderTool.handle()));
  ASG_CHECK(m_reclusterJetTool.setProperty("JetModifiers", modArray));
  ASG_CHECK(m_reclusterJetTool.setProperty("OutputLevel", msg().level() ) );
  ASG_CHECK(m_reclusterJetTool.retrieve());

  // clear modArray again
  modArray.clear();
  //        then trim the reclustered jets
  ATH_MSG_INFO( "Jet Trimmer initializing..." );
  ATH_MSG_INFO( "\tPtFrac: " << m_ptFrac );
  ATH_MSG_INFO( "\tRClus: " << m_subjet_radius );
  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_jetTrimmingTool, JetTrimmer) );
  ASG_CHECK(m_jetTrimmingTool.setProperty("PtFrac", m_ptFrac));
  ASG_CHECK(m_jetTrimmingTool.setProperty("RClus", m_subjet_radius));
  ASG_CHECK(m_jetTrimmingTool.setProperty("JetBuilder", m_jetFromPseudoJetTool.handle()));
  ASG_CHECK(m_jetTrimmingTool.setProperty("OutputLevel", msg().level() ) );
  ASG_CHECK(m_jetTrimmingTool.retrieve() );
  //        and then apply all other modifiers based on the trimmed reclustered jets
  ATH_MSG_INFO( "\t... and queuing up various jet modifiers..." );

  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_jetChargeTool, JetChargeTool) );
  ASG_CHECK(m_jetChargeTool.retrieve());
  modArray.push_back(m_jetChargeTool.handle() );

  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_jetPullTool, JetPullTool) );
  ASG_CHECK(m_jetPullTool.retrieve());
  modArray.push_back(m_jetPullTool.handle() );

  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_energyCorrelatorTool, EnergyCorrelatorTool) );
  ASG_CHECK(m_energyCorrelatorTool.retrieve());
  modArray.push_back(m_energyCorrelatorTool.handle() );

  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_energyCorrelatorRatiosTool, EnergyCorrelatorRatiosTool) );
  ASG_CHECK(m_energyCorrelatorRatiosTool.retrieve());
  modArray.push_back(m_energyCorrelatorRatiosTool.handle() );

  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_ktSplittingScaleTool, KTSplittingScaleTool) );
  ASG_CHECK(m_ktSplittingScaleTool.retrieve());
  modArray.push_back(m_ktSplittingScaleTool.handle() );

  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_dipolarityTool, DipolarityTool) );
  ASG_CHECK(m_dipolarityTool.retrieve());
  modArray.push_back(m_dipolarityTool.handle() );

  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_centerOfMassShapesTool, CenterOfMassShapesTool) );
  ASG_CHECK(m_centerOfMassShapesTool.retrieve());
  modArray.push_back(m_centerOfMassShapesTool.handle() );

  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_nSubjettinessTool, NSubjettinessTool) );
  ASG_CHECK(m_nSubjettinessTool.retrieve());
  modArray.push_back(m_nSubjettinessTool.handle() );
  ATH_CHECK(modArray.retrieve() );
  // finish up the rest of the tool

  ASG_CHECK( ASG_MAKE_ANA_TOOL( m_trimJetTool, JetRecTool) );
  ASG_CHECK(m_trimJetTool.setProperty("InputContainer", reclusteredJetsContainer));
  ASG_CHECK(m_trimJetTool.setProperty("OutputContainer", m_outputJetContainer));
  ASG_CHECK(m_trimJetTool.setProperty("JetModifiers", modArray));
  ASG_CHECK(m_trimJetTool.setProperty("JetGroomer",  m_jetTrimmingTool.handle() ));
  ASG_CHECK(m_trimJetTool.setProperty("OutputLevel", msg().level() ) );
  ASG_CHECK(m_trimJetTool.retrieve());

  return StatusCode::SUCCESS;
}

int JetReclusteringTool::execute() const {
  int retCode = 0;
  if(m_ptMin_input > 0) {
    if ( (retCode = m_inputJetFilterTool->execute() ) != 0) {
      ATH_MSG_ERROR( "Failure in inputJetFilterTool with return code: " << retCode );
      return retCode;
    }
  }
  if ( (retCode = m_reclusterJetTool->execute() ) != 0) {
    ATH_MSG_ERROR( "Failure in reclusterJetTool with return code: " << retCode );
    return retCode;
  }
  if ( (retCode = m_trimJetTool->execute() ) != 0) {
    ATH_MSG_ERROR( "Failure in trimJetTool with return code: " << retCode );
    return retCode;
  }
  return 0;
}

void JetReclusteringTool::print() const {
  ATH_MSG_INFO("Properties for " << m_APP_NAME << "(" << name() << ")");
  ATH_MSG_INFO("    InputJetContainer:      " << m_inputJetContainer );
  ATH_MSG_INFO("    OutputJetContainer:     " << m_outputJetContainer );
  ATH_MSG_INFO("    Radius:                 " << m_radius );
  ATH_MSG_INFO("    ReclusteringAlgorithm:  " << m_rc_alg << " (" << m_rc_alg << ")" );
  ATH_MSG_INFO("    VariableRMinRadius:     " << m_varR_minR );
  ATH_MSG_INFO("    VariableRMassScale:     " << m_varR_mass << " GeV" );
  ATH_MSG_INFO("    InputJetPtCut:          " << m_ptMin_input << " GeV" );
  ATH_MSG_INFO("    ReclusteredJetPtCut:    " << m_ptMin_rc << " GeV" );
  ATH_MSG_INFO("    ReclusteredJetPtFrac:   " << m_ptFrac );
  ATH_MSG_INFO("    ReclusteredJetSubjetR:  " << m_subjet_radius );

  if(m_isInitialized){
    m_inputJetFilterTool->print();
    m_reclusterJetTool->print();
    m_trimJetTool->print();
  } else {
    ATH_MSG_INFO(m_APP_NAME << " has not been initialized yet");
  }

  return;
}
